BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -84,13 +84,16 @@ def new #:nodoc:
                                 )
 
     @typelist = get_type_list
+    render template: 'data_providers/normal_new' unless current_user.has_role?(:admin_user) # normal user only allowed create UserkeyFlatDirSshDataProvider
   end
 
   def create #:nodoc:
-    @provider = DataProvider.sti_new(data_provider_params)
+    @provider = DataProvider.sti_new(data_provider_params) if current_user.has_role?(:admin_user)
```

Comment:
You modified the create action to add support for normal users creating their own DP. The create action used to be restricted to admins, so because of that it was allowed to do a lot of things without checking for security issues. Now that normal users can invoke it, they can trigger these security issues. E.g. they can create a DP for another user or for a project that they don't have access!

Instead of trying to retrofit this original create(), set up a separate action for creating the user DPs, a bit like it was done on the neurohub side. That will also provide you with examples of what a user is allowed to set or not in the object.

Reply:
The new action could be called "create_userdp" or something.